docs/optimization.md

```diff
@@ -0,0 +1,153 @@
+    - [ ]  Add information for subquery transformation above
+- [ ]  Add information about aggregations in `SELECT` here
+- [ ]  Add information about `SelectDistinct` in `SELECT` here
+- If a `LIMIT` exists in the select statement, the value of the limit and associated offset, along with any  `ORDER_BY` clauses, are stored in a `LogicalLimit` node that is added as the parent of the existing `output_expr_`
+
+Our example query has both a `LogicalLimit` and set of `LogicalFilter`s above a `LogicalGet` in the `output_expr_`.
+
+Following the above transformations, we create a `QueryInfo` object that keeps track of the query type. Additionally, on a `SELECT` statement, we also keep track of the output columns and sort properties in the `QueryInfo` object.
+
+The top-level function call of the following sections is `Optimizer::BuildPlanTree`, which takes as input:
+
+- `transaction::TransactionContext *txn`: the active transaction
+- `catalog::CatalogAccessor *accessor`: the catalog accessor
+- `StatsStorage *storage`: a structure to manage table statistics
+- `QueryInfo query_info`: information directly relevant to active query (type, output expressions, physical properties)
+- `std::unique_ptr<AbstractOptimizerNode> op_tree`: the logical expression tree generated from the transformation
+
+and will eventually output the physical plan chosen.
+
+# 3. Optimization
+
+The bulk of the optimization logic occurs in this section. The optimization loop generates all transformations, . Some key related definitions follow:
+
+- `OptimizationContext`: the structure which keeps track of required properties, costs, and the OptimizerContext for each optimization process.
+- `OptimizerContext`: the structure which tracks the cost model, task pool, transaction context, statistics storage, catalog accessor, rule set, and tracked optimization contexts, and importantly the **memoized expressions for each group.**
+- `Memo`: a class to track groups and group expressions that allows for duplicate detection, rewrites, access, and erase procedures.
+- `GroupExpression`: a class to track a specific operator (logical or physical) in a group which tracks the transaction, group number, whether the stats have been derived, the lowest cost satisfying the required properties of an expression, and the details of the expression.
+- `Group`: a class to collect group expressions representing logically equivalent expressions including logical and physical transformations.
+- `LogicalExpression`: a group expression which represents a logical expression i.e. the logical operator that specifies the general behavior of a node
+- `PhysicalExpression`: a group expression which represents a logical expression i.e. the physical operator that specifies the implementation details of the expression
```

Comment:
Did you mean to say "logical expression" here?

Reply:
Which line is this about?

Reply:
"`PhysicalExpression`: a group expression which represents a logical expression..." I think you meant to put "physical expression" instead of "logical expression".